File: go/types/path_test.go
// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

package types

import (
	"fmt"
	"testing"

	"bytes"

	"github.com/attic-labs/noms/go/hash"
	"github.com/attic-labs/testify/assert"
)

func hashIdx(v Value) string {
	return fmt.Sprintf("[#%s]", v.Hash().String())
}

func assertResolvesTo(assert *assert.Assertions, expect, ref Value, str string) {
	p, err := ParsePath(str)
	assert.NoError(err)
	actual := p.Resolve(ref)
	if expect == nil {
		if actual != nil {
			assert.Fail("", "Expected nil, but got %s", EncodedValue(actual))
		}
	} else if actual == nil {
		assert.Fail("", "Expected %s, but got nil", EncodedValue(expect))
	} else {
		assert.True(expect.Equals(actual), "Expected %s, but got %s", EncodedValue(expect), EncodedValue(actual))
	}
}

func TestPathStruct(t *testing.T) {
	assert := assert.New(t)

	v := NewStruct("", StructData{
		"foo": String("foo"),
		"bar": Bool(false),
		"baz": Number(203),
	})

	assertResolvesTo(assert, String("foo"), v, `.foo`)
	assertResolvesTo(assert, Bool(false), v, `.bar`)
	assertResolvesTo(assert, Number(203), v, `.baz`)
	assertResolvesTo(assert, nil, v, `.notHere`)

	v2 := NewStruct("", StructData{
		"v1": v,
	})

	assertResolvesTo(assert, String("foo"), v2, `.v1.foo`)
	assertResolvesTo(assert, Bool(false), v2, `.v1.bar`)
	assertResolvesTo(assert, Number(203), v2, `.v1.baz`)
	assertResolvesTo(assert, nil, v2, `.v1.notHere`)
	assertResolvesTo(assert, nil, v2, `.notHere.v1`)
}

func TestPathIndex(t *testing.T) {
	assert := assert.New(t)

	var v Value
	resolvesTo := func(expVal, expKey Value, str string) {
		assertResolvesTo(assert, expVal, v, str)
		assertResolvesTo(assert, expKey, v, str+"@key")
	}

	v = NewList(Number(1), Number(3), String("foo"), Bool(false))

	resolvesTo(Number(1), Number(0), "[0]")
	resolvesTo(Number(3), Number(1), "[1]")
	resolvesTo(String("foo"), Number(2), "[2]")
	resolvesTo(Bool(false), Number(3), "[3]")
	resolvesTo(nil, nil, "[4]")
	resolvesTo(nil, nil, "[-5]")
	resolvesTo(Number(1), Number(0), "[-4]")
	resolvesTo(Number(3), Number(1), "[-3]")
	resolvesTo(String("foo"), Number(2), "[-2]")
	resolvesTo(Bool(false), Number(3), "[-1]")

	v = NewMap(
		Bool(false), Number(23),
		Number(1), String("foo"),
		Number(2.3), Number(4.5),
		String("two"), String("bar"),
	)

	resolvesTo(String("foo"), Number(1), "[1]")
	resolvesTo(String("bar"), String("two"), `["two"]`)
	resolvesTo(Number(23), Bool(false), "[false]")
	resolvesTo(Number(4.5), Number(2.3), "[2.3]")
	resolvesTo(nil, nil, "[4]")
}

func TestPathHashIndex(t *testing.T) {
	assert := assert.New(t)

	b := Bool(true)
	br := NewRef(b)
	i := Number(0)
	str := String("foo")
	l := NewList(b, i, str)
	lr := NewRef(l)
	m := NewMap(
		b, br,
		br, i,
		i, str,
		l, lr,
		lr, b,
	)
	s := NewSet(b, br, i, str, l, lr)

	resolvesTo := func(col, key, expVal, expKey Value) {
		assertResolvesTo(assert, expVal, col, hashIdx(key))
		assertResolvesTo(assert, expKey, col, hashIdx(key)+"@key")
	}

	// Primitives are only addressable by their values.
	resolvesTo(m, b, nil, nil)
	resolvesTo(m, i, nil, nil)
	resolvesTo(m, str, nil, nil)
	resolvesTo(s, b, nil, nil)
	resolvesTo(s, i, nil, nil)
	resolvesTo(s, str, nil, nil)

	// Other values are only addressable by their hashes.
	resolvesTo(m, br, i, br)
	resolvesTo(m, l, lr, l)
	resolvesTo(m, lr, b, lr)
	resolvesTo(s, br, br, br)
	resolvesTo(s, l, l, l)
	resolvesTo(s, lr, lr, lr)

	// Lists cannot be addressed by hashes, obviously.
	resolvesTo(l, i, nil, nil)
}

func TestPathHashIndexOfSingletonCollection(t *testing.T) {
	// This test is to make sure we don't accidentally return |b| if it's the only element.
	assert := assert.New(t)

	resolvesToNil := func(col, val Value) {
		assertResolvesTo(assert, nil, col, hashIdx(val))
	}

	b := Bool(true)
	resolvesToNil(NewMap(b, b), b)
	resolvesToNil(NewSet(b), b)
}

func TestPathMulti(t *testing.T) {
	assert := assert.New(t)

	m1 := NewMap(
		String("a"), String("foo"),
		String("b"), String("bar"),
		String("c"), String("car"),
	)

	m2 := NewMap(
		Bool(false), String("earth"),
		String("d"), String("dar"),
		m1, String("fire"),
	)

	l := NewList(m1, m2)

	s := NewStruct("", StructData{
		"foo": l,
	})

	assertResolvesTo(assert, l, s, `.foo`)
	assertResolvesTo(assert, m1, s, `.foo[0]`)
	assertResolvesTo(assert, String("foo"), s, `.foo[0]["a"]`)
	assertResolvesTo(assert, String("bar"), s, `.foo[0]["b"]`)
	assertResolvesTo(assert, String("car"), s, `.foo[0]["c"]`)
	assertResolvesTo(assert, String("foo"), s, `.foo[0]@at(0)`)
	assertResolvesTo(assert, String("bar"), s, `.foo[0]@at(1)`)
	assertResolvesTo(assert, String("car"), s, `.foo[0]@at(2)`)
	assertResolvesTo(assert, nil, s, `.foo[0]["x"]`)
	assertResolvesTo(assert, nil, s, `.foo[2]["c"]`)
	assertResolvesTo(assert, nil, s, `.notHere[0]["c"]`)
	assertResolvesTo(assert, m2, s, `.foo[1]`)
	assertResolvesTo(assert, String("dar"), s, `.foo[1]["d"]`)
	assertResolvesTo(assert, String("earth"), s, `.foo[1][false]`)
	assertResolvesTo(assert, String("fire"), s, fmt.Sprintf(`.foo[1]%s`, hashIdx(m1)))
	assertResolvesTo(assert, m1, s, fmt.Sprintf(`.foo[1]%s@key`, hashIdx(m1)))
	assertResolvesTo(assert, String("car"), s, fmt.Sprintf(`.foo[1]%s@key["c"]`, hashIdx(m1)))
	assertResolvesTo(assert, String("fire"), s, `.foo[1]@at(2)`)
	assertResolvesTo(assert, m1, s, `.foo[1]@at(2)@key`)
	assertResolvesTo(assert, String("car"), s, `.foo[1]@at(2)@key@at(2)`)
	assertResolvesTo(assert, String("fire"), s, `.foo[1]@at(-1)`)
	assertResolvesTo(assert, m1, s, `.foo[1]@at(-1)@key`)
	assertResolvesTo(assert, String("car"), s, `.foo[1]@at(-1)@key@at(-1)`)
}

func TestPathParseSuccess(t *testing.T) {
	assert := assert.New(t)

	test := func(str string) {
		p, err := ParsePath(str)
		assert.NoError(err)
		expectStr := str
		switch expectStr { // Human readable serialization special cases.
		case "[1e4]":
			expectStr = "[10000]"
		case "[1.]":
			expectStr = "[1]"
		case "[\"line\nbreak\rreturn\"]":
			expectStr = `["line\nbreak\rreturn"]`
		}
		assert.Equal(expectStr, p.String())
	}

	h := Number(42).Hash() // arbitrary hash

	test(".foo")
	test(".foo@type")
	test(".Q")
	test(".QQ")
	test("[true]")
	test("[true]@type")
	test("[false]")
	test("[false]@key")
	test("[false]@key@type")
	test("[false]@key@type@at(42)")
	test("[42]")
	test("[42]@key")
	test("[42]@at(-101)")
	test("[1e4]")
	test("[1.]")
	test("[1.345]")
	test(`[""]`)
	test(`["42"]`)
	test(`["42"]@key`)
	test("[\"line\nbreak\rreturn\"]")
	test(`["qu\\ote\""]`)
	test(`["π"]`)
	test(`["[[br][]acke]]ts"]`)
	test(`["xπy✌z"]`)
	test(`["ಠ_ಠ"]`)
	test(`["0"]["1"]["100"]`)
	test(".foo[0].bar[4.5][false]")
	test(fmt.Sprintf(".foo[#%s]", h.String()))
	test(fmt.Sprintf(".bar[#%s]@key", h.String()))
}

func TestPathParseErrors(t *testing.T) {
	assert := assert.New(t)

	test := func(str, expectError string) {
		p, err := ParsePath(str)
		assert.Equal(Path{}, p)
		if err != nil {
			assert.Equal(expectError, err.Error())
		} else {
			assert.Fail("Expected " + expectError)
		}
	}

	test("", "Empty path")
	test(".", "Invalid field: ")
	test("[", "Path ends in [")
	test("]", "] is missing opening [")
	test(".#", "Invalid field: #")
	test(". ", "Invalid field:  ")
	test(". invalid.field", "Invalid field:  invalid.field")
	test(".foo.", "Invalid field: ")
	test(".foo.#invalid.field", "Invalid field: #invalid.field")
	test(".foo!", "Invalid operator: !")
	test(".foo!bar", "Invalid operator: !")
	test(".foo#", "Invalid operator: #")
	test(".foo#bar", "Invalid operator: #")
	test(".foo[", "Path ends in [")
	test(".foo[.bar", "Invalid index: .bar")
	test(".foo]", "] is missing opening [")
	test(".foo].bar", "] is missing opening [")
	test(".foo[]", "Empty index value")
	test(".foo[[]", "Invalid index: [")
	test(".foo[[]]", "Invalid index: [")
	test(".foo[42.1.2]", "Invalid index: 42.1.2")
	test(".foo[1f4]", "Invalid index: 1f4")
	test(".foo[hello]", "Invalid index: hello")
	test(".foo['hello']", "Invalid index: 'hello'")
	test(`.foo[\]`, `Invalid index: \`)
	test(`.foo[\\]`, `Invalid index: \\`)
	test(`.foo["hello]`, "[ is missing closing ]")
	test(`.foo["hello`, "[ is missing closing ]")
	test(`.foo["hello"`, "[ is missing closing ]")
	test(`.foo["`, "[ is missing closing ]")
	test(`.foo["\`, "[ is missing closing ]")
	test(`.foo["]`, "[ is missing closing ]")
	test(".foo[#]", "Invalid hash: ")
	test(".foo[#invalid]", "Invalid hash: invalid")
	test(`.foo["hello\nworld"]`, `Only " and \ can be escaped`)
	test(".foo[42]bar", "Invalid operator: b")
	test("#foo", "Invalid operator: #")
	test("!foo", "Invalid operator: !")
	test("@foo", "Unsupported annotation: @foo")
	test("@key", "Cannot use @key annotation at beginning of path")
	test(".foo@key", "Cannot use @key annotation on: .foo")
	test(".foo@key()", "@key annotation does not support arguments")
	test(".foo@key(42)", "@key annotation does not support arguments")
	test(".foo@type()", "@type annotation does not support arguments")
	test(".foo@type(42)", "@type annotation does not support arguments")
	test(".foo@at", "@at annotation requires a position argument")
	test(".foo@at()", "@at annotation requires a position argument")
	test(fmt.Sprintf(".foo[#%s]@soup", hash.Of([]byte{42}).String()), "Unsupported annotation: @soup")
}

func TestPathEquals(t *testing.T) {
	assert := assert.New(t)
	equalPaths := []string{
		`[1]`,
		`["one"]`,
		`.two.three`,
		`["yo"]@key`,
	}
	notEqualPaths := [][]string{
		{`[1]`, `[2]`},
		{`["one"]`, `["two"]`},
		{`.two.three`, `.two.four`},
		{`["yo"]@key`, `["yo"]`},
	}

	assert.True(Path{}.Equals(Path{}))
	for _, s := range equalPaths {
		p, err := ParsePath(s)
		assert.NoError(err)
		assert.True(p.Equals(p))
	}

	simple, err := ParsePath(`["one"].two`)
	assert.NoError(err)
	assert.False(Path{}.Equals(simple))
	for _, a := range notEqualPaths {
		s0, s1 := a[0], a[1]
		p0, err := ParsePath(s0)
		assert.NoError(err)
		p1, err := ParsePath(s1)
		assert.NoError(err)
		assert.False(p0.Equals(p1))
	}
}

func TestPathCanBePathIndex(t *testing.T) {
	assert := assert.New(t)

	assert.True(ValueCanBePathIndex(Bool(true)))
	assert.True(ValueCanBePathIndex(Number(5)))
	assert.True(ValueCanBePathIndex(String("yes")))

	assert.False(ValueCanBePathIndex(NewRef(String("yes"))))
	assert.False(ValueCanBePathIndex(NewBlob(bytes.NewReader([]byte("yes")))))
}

func TestCopyPath(t *testing.T) {
	assert := assert.New(t)

	testCases := []string{
		``,
		`["key"]`,
		`["key"].field1`,
		`["key"]@key.field1`,
	}

	for _, s1 := range testCases {
		expected, err := ParsePath(s1 + `["anIndex"]`)
		assert.NoError(err)
		var p Path
		if s1 != "" {
			p, err = ParsePath(s1)
		}
		assert.NoError(err)
		p1 := p.Append(NewIndexPath(String("anIndex")))
		if len(p) > 0 {
			p[0] = expected[1] // if p1 really is a copy, this shouldn't be noticed
		}
		assert.Equal(expected, p1)
	}
}

func TestMustParsePath(t *testing.T) {
	for _, good := range []string{".good", "[\"good\"]"} {
		assert.NotNil(t, MustParsePath(good))
	}
	for _, bad := range []string{"", "bad", "[bad]", "!", "💩"} {
		assert.Panics(t, func() { MustParsePath(bad) })
	}
}

func TestPathType(t *testing.T) {
	assert := assert.New(t)

	m := NewMap(
		String("string"), String("foo"),
		String("bool"), Bool(false),
		String("number"), Number(42),
		String("List<number|string>"), NewList(Number(42), String("foo")),
		String("Map<Bool, Bool>"), NewMap(Bool(true), Bool(false)))

	m.IterAll(func(k, cv Value) {
		ks := k.(String)
		assertResolvesTo(assert, cv.Type(), m, fmt.Sprintf("[\"%s\"]@type", ks))
	})

	assertResolvesTo(assert, StringType, m, `["string"]@key@type`)
	assertResolvesTo(assert, m.Type(), m, `@type`)
	s := NewStruct("", StructData{
		"str": String("foo"),
		"num": Number(42),
	})
	assertResolvesTo(assert, s.Get("str").Type(), s, ".str@type")
	assertResolvesTo(assert, s.Get("num").Type(), s, ".num@type")
}

func TestPathAtAnnotation(t *testing.T) {
	assert := assert.New(t)

	var v Value
	resolvesTo := func(expVal, expKey Value, str string) {
		assertResolvesTo(assert, expVal, v, str)
		assertResolvesTo(assert, expKey, v, str+"@key")
	}

	v = NewList(Number(1), Number(3), String("foo"), Bool(false))

	resolvesTo(Number(1), nil, "@at(0)")
	resolvesTo(Number(3), nil, "@at(1)")
	resolvesTo(String("foo"), nil, "@at(2)")
	resolvesTo(Bool(false), nil, "@at(3)")
	resolvesTo(nil, nil, "@at(4)")
	resolvesTo(nil, nil, "@at(-5)")
	resolvesTo(Number(1), nil, "@at(-4)")
	resolvesTo(Number(3), nil, "@at(-3)")
	resolvesTo(String("foo"), nil, "@at(-2)")
	resolvesTo(Bool(false), nil, "@at(-1)")

	v = NewSet(
		Bool(false),
		Number(1),
		Number(2.3),
		String("two"),
	)

	resolvesTo(Bool(false), Bool(false), "@at(0)")
	resolvesTo(Number(1), Number(1), "@at(1)")
	resolvesTo(Number(2.3), Number(2.3), "@at(2)")
	resolvesTo(String("two"), String("two"), `@at(3)`)
	resolvesTo(nil, nil, "@at(4)")
	resolvesTo(nil, nil, "@at(-5)")
	resolvesTo(Bool(false), Bool(false), "@at(-4)")
	resolvesTo(Number(1), Number(1), "@at(-3)")
	resolvesTo(Number(2.3), Number(2.3), "@at(-2)")
	resolvesTo(String("two"), String("two"), `@at(-1)`)

	v = NewMap(
		Bool(false), Number(23),
		Number(1), String("foo"),
		Number(2.3), Number(4.5),
		String("two"), String("bar"),
	)

	resolvesTo(Number(23), Bool(false), "@at(0)")
	resolvesTo(String("foo"), Number(1), "@at(1)")
	resolvesTo(Number(4.5), Number(2.3), "@at(2)")
	resolvesTo(String("bar"), String("two"), `@at(3)`)
	resolvesTo(nil, nil, "@at(4)")
	resolvesTo(nil, nil, "@at(-5)")
	resolvesTo(Number(23), Bool(false), "@at(-4)")
	resolvesTo(String("foo"), Number(1), "@at(-3)")
	resolvesTo(Number(4.5), Number(2.3), "@at(-2)")
	resolvesTo(String("bar"), String("two"), `@at(-1)`)
}

File: js/noms/src/path-test.js
// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

// @flow

import {assert} from 'chai';
import {suite, test} from './jest.js';
import {equals} from './compare.js';

import {getHash} from './get-hash.js';
import {notNull} from './assert.js';
import List from './list.js';
import Map from './map.js';
import Path from './path.js';
import Ref from './ref.js';
import Set from './set.js';
import {getTypeOfValue, numberType, stringType} from './type.js';
import type Value from './value.js';
import {newStruct} from './struct.js';

function hashIdx(v: Value): string {
  return `[#${getHash(v).toString()}]`;
}

async function assertResolvesTo(expect: Value | null, target: Value, str: string) {
  const j = s => JSON.stringify(s);
  const p = Path.parse(str);
  const actual = await p.resolve(target);
  if (expect === null) {
    assert.isTrue(actual === null, `Expected null, but got ${j(actual)}`);
  } else if (actual === null) {
    assert.isTrue(false, `Expected ${j(expect)}, but got null`);
  } else {
    assert.isTrue(equals(notNull(expect), actual), `Expected ${j(expect)}, but got ${j(actual)}`);
  }
}

suite('Path', () => {
  test('struct', async () => {
    const v = newStruct('', {
      foo: 'foo',
      bar: false,
      baz: 203,
    });

    await assertResolvesTo('foo', v, '.foo');
    await assertResolvesTo(false, v, '.bar');
    await assertResolvesTo(203, v, '.baz');
    await assertResolvesTo(null, v, '.notHere');

    const v2 = newStruct('', {
      v1: v,
    });

    await assertResolvesTo('foo', v2, '.v1.foo');
    await assertResolvesTo(false, v2, '.v1.bar');
    await assertResolvesTo(203, v2, '.v1.baz');
    await assertResolvesTo(null, v2, '.v1.notHere');
    await assertResolvesTo(null, v2, '.notHere.foo');
  });

  test('index', async () => {
    let v: Value;
    const resolvesTo = async (exp: Value | null, val: Value, str: string) => {
      // Indices resolve to |exp|.
      await assertResolvesTo(exp, v, str);
      // Keys resolves to themselves.
      if (exp !== null) {
        exp = val;
      }
      await assertResolvesTo(exp, v, str + '@key');
    };

    v = new List([1, 3, 'foo', false]);

    await resolvesTo(1, 0, '[0]');
    await resolvesTo(3, 1, '[1]');
    await resolvesTo('foo', 2, '[2]');
    await resolvesTo(false, 3, '[3]');
    await resolvesTo(null, 4, '[4]');
    await resolvesTo(null, 4, '[-5]');
    await resolvesTo(1, 0, '[-4]');
    await resolvesTo(3, 1, '[-3]');
    await resolvesTo('foo', 2, '[-2]');
    await resolvesTo(false, 3, '[-1]');

    v = new Map([
      [1, 'foo'],
      ['two', 'bar'],
      [false, 23],
      [2.3, 4.5],
    ]);

    await resolvesTo('foo', 1, '[1]');
    await resolvesTo('bar', 'two', '["two"]');
    await resolvesTo(23, false, '[false]');
    await resolvesTo(4.5, 2.3, '[2.3]');
    await resolvesTo(null, 4, '[4]');
  });

  test('hash index', async () => {
    const b = true;
    const br = new Ref(b);
    const i = 0;
    const str = 'foo';
    const l = new List([b, i, str]);
    const lr = new Ref(l);
    const m = new Map([
      [b, br],
      [br, i],
      [i, str],
      [l, lr],
      [lr, b],
    ]);
    const s = new Set([b, br, i, str, l, lr]);

    const resolvesTo = async (col: Value, exp: Value | null, val: Value) => {
      // Values resolve to |exp|.
      await assertResolvesTo(exp, col, hashIdx(val));
      // Keys resolves to themselves.
      if (exp !== null) {
        exp = val;
      }
      await assertResolvesTo(exp, col, hashIdx(val) + '@key');
    };

    // Primitives are only addressable by their values.
    await resolvesTo(m, null, b);
    await resolvesTo(m, null, i);
    await resolvesTo(m, null, str);
    await resolvesTo(s, null, b);
    await resolvesTo(s, null, i);
    await resolvesTo(s, null, str);

    // Other values are only addressable by their hashes.
    await resolvesTo(m, i, br);
    await resolvesTo(m, lr, l);
    await resolvesTo(m, b, lr);
    await resolvesTo(s, br, br);
    await resolvesTo(s, l, l);
    await resolvesTo(s, lr, lr);

    // Lists cannot be addressed by hashes, obviously.
    await resolvesTo(l, null, i);
  });

  test('hash index of singleton collection', async () => {
    // This test is to make sure we don't accidentally return the element of a singleton map.
    const resolvesToNull = async (col: Value, v: Value) => {
      await assertResolvesTo(null, col, hashIdx(v));
    };

    await resolvesToNull(new Map([[true, true]]), true);
    await resolvesToNull(new Set([true]), true);
  });

  test('multi', async () => {
    const m1 = new Map([
      ['a', 'foo'],
      ['b', 'bar'],
      ['c', 'car'],
    ]);

    const m2 = new Map([
      ['d', 'dar'],
      [false, 'earth'],
      [m1, 'fire'],
    ]);

    const l = new List([m1, m2]);

    const s = newStruct('', {
      'foo': l,
    });

    await assertResolvesTo(l, s, '.foo');
    await assertResolvesTo(m1, s, '.foo[0]');
    await assertResolvesTo('foo', s, '.foo[0]["a"]');
    await assertResolvesTo('bar', s, '.foo[0]["b"]');
    await assertResolvesTo('car', s, '.foo[0]["c"]');
    await assertResolvesTo('foo', s, '.foo[0]@at(0)');
    await assertResolvesTo('bar', s, '.foo[0]@at(1)');
    await assertResolvesTo('car', s, '.foo[0]@at(2)');
    await assertResolvesTo(null, s, '.foo[0]["x"]');
    await assertResolvesTo(null, s, '.foo[0]@at(3)');
    await assertResolvesTo(null, s, '.foo[2]["c"]');
    await assertResolvesTo(null, s, '.notHere[0]["c"]');
    await assertResolvesTo(m2, s, '.foo[1]');
    await assertResolvesTo('dar', s, '.foo[1]["d"]');
    await assertResolvesTo('earth', s, '.foo[1][false]');
    await assertResolvesTo('fire', s, `.foo[1]${hashIdx(m1)}`);
    await assertResolvesTo(m1, s, `.foo[1]${hashIdx(m1)}@key`);
    await assertResolvesTo('car', s, `.foo[1]${hashIdx(m1)}@key["c"]`);
    await assertResolvesTo('fire', s, '.foo[1]@at(2)');
    await assertResolvesTo(m1, s, '.foo[1]@at(2)@key');
    await assertResolvesTo('car', s, '.foo[1]@at(2)@key@at(2)');
    await assertResolvesTo('fire', s, '.foo[1]@at(-1)');
    await assertResolvesTo(m1, s, '.foo[1]@at(-1)@key');
    await assertResolvesTo('car', s, '.foo[1]@at(-1)@key@at(-1)');
  });

  test('parse success', () => {
    const t = (s: string) => {
      const p = Path.parse(s);
      let expect = s;
      // Human readable serialization special cases.
      if (expect === '[1e4]') {
        expect = '[10000]';
      } else if (expect === '[1.]') {
        expect = '[1]';
      } else if (expect === '["line\nbreak\rreturn"]') {
        expect = '["line\\nbreak\\rreturn"]';
      }
      assert.strictEqual(expect, p.toString());
    };

    const h = getHash(42); // arbitrary hash

    t('.foo');
    t('.foo@type');
    t('.Q');
    t('.QQ');
    t('[true]');
    t('[true]@type');
    t('[false]');
    t('[false]@key');
    t('[false]@key@type');
    t('[false]@key@type@at(42)');
    t('[42]');
    t('[42]@key');
    t('[42]@at(101)');
    t('[1e4]');
    t('[1.]');
    t('[1.345]');
    t('[""]');
    t('["42"]');
    t('["42"]@key');
    t('[\"line\nbreak\rreturn\"]');
    t('["qu\\\\ote\\\""]');
    t('["π"]');
    t('["[[br][]acke]]ts"]');
    t('["xπy✌z"]');
    t('["ಠ_ಠ"]');
    t('["0"]["1"]["100"]');
    t('.foo[0].bar[4.5][false]');
    t(`.foo[#${h.toString()}]`);
    t(`.bar[#${h.toString()}]@key`);
  });

  test('parse errors', () => {
    const t = (s: string, expectErr: string) => {
      let actualErr = '';
      try {
        Path.parse(s);
      } catch (e) {
        assert.instanceOf(e, SyntaxError);
        actualErr = e.message;
      }
      assert.strictEqual(expectErr, actualErr);
    };

    t('', 'Empty path');
    t('.', 'Invalid field: ');
    t('[', 'Path ends in [');
    t(']', '] is missing opening [');
    t('.#', 'Invalid field: #');
    t('. ', 'Invalid field:  ');
    t('. invalid.field', 'Invalid field:  invalid.field');
    t('.foo.', 'Invalid field: ');
    t('.foo.#invalid.field', 'Invalid field: #invalid.field');
    t('.foo!', 'Invalid operator: !');
    t('.foo!bar', 'Invalid operator: !');
    t('.foo#', 'Invalid operator: #');
    t('.foo#bar', 'Invalid operator: #');
    t('.foo[', 'Path ends in [');
    t('.foo[.bar', '[ is missing closing ]');
    t('.foo]', '] is missing opening [');
    t('.foo].bar', '] is missing opening [');
    t('.foo[]', 'Empty index value');
    t('.foo[[]', 'Invalid index: [');
    t('.foo[[]]', 'Invalid index: [');
    t('.foo[42.1.2]', 'Invalid index: 42.1.2');
    t('.foo[1f4]', 'Invalid index: 1f4');
    t('.foo[hello]', 'Invalid index: hello');
    t('.foo[\'hello\']', 'Invalid index: \'hello\'');
    t('.foo[\\]', 'Invalid index: \\');
    t('.foo[\\\\]', 'Invalid index: \\\\');
    t('.foo["hello]', '[ is missing closing ]');
    t('.foo["hello', '[ is missing closing ]');
    t('.foo["', '[ is missing closing ]');
    t('.foo["\\', '[ is missing closing ]');
    t('.foo["]', '[ is missing closing ]');
    t('.foo[#]', 'Invalid hash: ');
    t('.foo[#invalid]', 'Invalid hash: invalid');
    t('.foo["hello\\nworld"]', 'Only " and \\ can be escaped');
    t('.foo[42]bar', 'Invalid operator: b');
    t('#foo', 'Invalid operator: #');
    t('!foo', 'Invalid operator: !');
    t('@foo', 'Unsupported annotation: @foo');
    t('@key', 'Cannot use @key annotation at beginning of path');
    t('.foo[42]@soup', 'Unsupported annotation: @soup');
    t('.foo@key', 'Cannot use @key annotation on: .foo');
    t('.foo@key()', '@key annotation does not support arguments');
    t('.foo@key(42)', '@key annotation does not support arguments');
    t('.foo@type()', '@type annotation does not support arguments');
    t('.foo@type(42)', '@type annotation does not support arguments');
    t('.foo@at', '@at annotation requires a position argument');
    t('.foo@at()', '@at annotation requires a position argument');
    t('.foo@at(', '@at annotation requires a position argument');
    t('.foo@at(42', '@at annotation requires a position argument');
  });

  test('type annotation', async () => {
    const mkv = [
      ['string', 'foo'],
      ['bool', false],
      ['number', 42],
      ['List<number|string>', new List([42, 'foo'])],
      ['Map<bool, bool>', new Map([[true, false]])],
    ];
    const m = new Map(mkv);
    const s = newStruct('', {
      str: 'foo',
      num: 42,
    });

    const tests = [];

    for (const [k, v] of mkv) {
      tests.push(assertResolvesTo(getTypeOfValue(v), m, `["${k}"]@type`));
      tests.push(assertResolvesTo(stringType, m, `["${k}"]@key@type`));
    }
    tests.push(
      assertResolvesTo(stringType, m, '["string"]@key@type'),
      assertResolvesTo(m.type, m, '@type'),
    );
    tests.push(
      assertResolvesTo(stringType, s, '.str@type'),
      assertResolvesTo(numberType, s, '.num@type'),
    );

    await Promise.all(tests);
  });

  test('at annotation', async () => {
    let v: Value;
    const resolvesTo = (expVal, expKey, str) => Promise.all([
      assertResolvesTo(expVal, v, str),
      assertResolvesTo(expKey, v, str + '@key'),
    ]);

    v = new List([1, 3, 'foo', false]);

    await Promise.all([
      resolvesTo(1, null, '@at(0)'),
      resolvesTo(3, null, '@at(1)'),
      resolvesTo('foo', null, '@at(2)'),
      resolvesTo(false, null, '@at(3)'),
      resolvesTo(null, null, '@at(4)'),
      resolvesTo(null, null, '@at(-5)'),
      resolvesTo(1, null, '@at(-4)'),
      resolvesTo(3, null, '@at(-3)'),
      resolvesTo('foo', null, '@at(-2)'),
      resolvesTo(false, null, '@at(-1)'),
    ]);

    v = new Set([false, 1, 2.3, 'two']);

    await Promise.all([
      resolvesTo(false, false, '@at(0)'),
      resolvesTo(1, 1, '@at(1)'),
      resolvesTo(2.3, 2.3, '@at(2)'),
      resolvesTo('two', 'two', '@at(3)'),
      resolvesTo(null, null, '@at(4)'),
      resolvesTo(null, null, '@at(-5)'),
      resolvesTo(false, false, '@at(-4)'),
      resolvesTo(1, 1, '@at(-3)'),
      resolvesTo(2.3, 2.3, '@at(-2)'),
      resolvesTo('two', 'two', '@at(-1)'),
    ]);

    v = new Map([
      [false, 23],
      [1, 'foo'],
      [2.3, 4.5],
      ['two', 'bar'],
    ]);

    await Promise.all([
      resolvesTo(23, false, '@at(0)'),
      resolvesTo('foo', 1, '@at(1)'),
      resolvesTo(4.5, 2.3, '@at(2)'),
      resolvesTo('bar', 'two', '@at(3)'),
      resolvesTo(null, null, '@at(4)'),
      resolvesTo(null, null, '@at(-5)'),
      resolvesTo(23, false, '@at(-4)'),
      resolvesTo('foo', 1, '@at(-3)'),
      resolvesTo(4.5, 2.3, '@at(-2)'),
      resolvesTo('bar', 'two', '@at(-1)'),
    ]);
  });
});
